### PR TITLE
Pull request for anjayluh to review

### PR DESCRIPTION
**Pull request by anjayluh**
Please review this pull request to continue with your application process

### CODE REVIEW

File: src/controllers/groupe.controller.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import { NextFunction, Request, Response } from 'express';
+import { Group } from '@prisma/client';
+import { CreateGroupDto } from '@dtos/group.dto';
+import groupService from '@services/groups.service';
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:4]: `groupService` is imported as a default but instantiated as new `groupService()`. By convention, classes should be PascalCase (`GroupService`) while instances should be camelCase. Using lowercase for classes is misleading.

File: src/controllers/groupe.controller.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import { NextFunction, Request, Response } from 'express';
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:1]: Good use of explicit imports for Express types.

File: src/controllers/groupe.controller.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import { Group } from '@prisma/client';
+import { CreateGroupDto } from '@dtos/group.dto';
+import groupService from '@services/groups.service';
+
+class GroupController {
+  public groupService = new groupService();
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:7]: - Related to the issue mentioned on line 4 above, `groupService` should be `GroupService`. Otherwise, it violates naming conventions and may confuse other contributors to the code.
- It is okay to create the instance at the controller level. Consider using dependency injection for testability.

File: src/controllers/groupe.controller.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+  public getGroups = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+    try {
+      const findAllGroupsData: Group[] = await this.groupService.findAllGroup();
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:11]: - I recommend using plural naming consistently (e.g., `findAllGroups`) to align with the semantics.
- Also, we do need to ensure file names follow the same convention—most are named group, but this one is `groupe`, which is inconsistent and can slow down onboarding and contributions.

File: src/controllers/groupe.controller.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+      res.status(200).json({ data: findAllGroupsData, message: 'findAll' });
+    } catch (error) {
+      next(error);
+    }
+  };
+
+  public getGroupById = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+    try {
+      const groupId = Number(req.params.id);
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:21]: There is no validation on req.params.id. In case it’s not a number e.g. `Number('abc')` returns `NaN`, which will pass into the service. You should guard against `NaN`. Using `parseInt` with validation will be clearer and handle the `NaN` error.

File: src/controllers/groupe.controller.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+      res.status(200).json({ data: findOneGroupData, message: 'findOne' });
+    } catch (error) {
+      next(error);
+    }
+  };
+
+  public createGroup = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+    try {
+      const groupData: CreateGroupDto = req.body;
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:32]: Ther is no validation pipeline here. DTO suggests you plan to use class-validator or similar. Without `validateOrReject`, this line simply casts req.body without real validation. In NestJS, data transfer objects are often combined with [class-validator](https://github.com/typestack/class-validator) to enforce input rules. But in this case, `groupData` is only being cast to `CreateGroupDto`—it doesn’t actually validate the request body. 

Suggestion

```
// Convert plain request body to an instance of the DTO
  const groupData = plainToInstance(CreateGroupDto, req.body);

  // Validate the DTO fields based on decorators in CreateGroupDto
  await validateOrReject(groupData);
```

File: src/controllers/groupe.controller.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      res.status(201).json({ data: createGroupData, message: 'created' });
+    } catch (error) {
+      next(error);
+    }
+  };
+
+  public updateGroup = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+    try {
+      const groupId = Number(req.params.id);
+      const groupData: CreateGroupDto = req.body;
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:44]: Same validation gap as in `createGroup` on line 32.

File: src/controllers/groupe.controller.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+      res.status(201).json({ data: createGroupData, message: 'created' });
+    } catch (error) {
+      next(error);
+    }
+  };
+
+  public updateGroup = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
+    try {
+      const groupId = Number(req.params.id);
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:43]: Similar comment as mentioned on line 21 concerning the need for validation for NaN.

File: src/controllers/groupe.controller.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      const deleteGroupData: Group = await this.groupService.deleteGroup(groupId);
+
+      res.status(200).json({ data: deleteGroupData, message: 'deleted' });
+    } catch (error) {
+      next(error);
+    }
+  };
+
+  public joinGroup = async (req, res, next) => {
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:64]: There is no try/catch or next(error) in here. This breaks consistency with the rest of the controller and bypasses error middleware.

File: src/controllers/groupe.controller.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    const groupId = Number(req.params.id);
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:65]: req.body.userId has no type safety. It should be explicitly typed (number) and validated. See comment on line 21

File: src/controllers/groupe.controller.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    const userId = req.body.userId;
+
+    const users = await this.groupService.joinGroup(userId, groupId);
+
+    res.status(200).json(users)
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:70]: The structure of your response payload is inconsistent. Most methods return a JSON object like:
`res.json({ data: ..., message: '...' })`

But this is returning just the raw users array: `res.json(users)`
 
Why this matters:

Consistent response formats make it easier for frontend developers or API consumers to handle responses predictably.

Inconsistencies can lead to extra error handling, bugs, or confusion when integrating with the API.

Suggestion: Wrap all responses in the same `{ data, message }` structure:

`res.status(200).json({ data: users, message: 'users joined' });`

File: src/controllers/groupe.controller.ts [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  public leaveGroup = async (req, res, next) => {
+    const groupId = Number(req.params.id);
+    const userId = req.body.userId;
+
+    const users = await this.groupService.leaveGroup(userId, groupId);
+
+    res.status(200).json(users)
```
Comment (anjayluh) [inline src/controllers/groupe.controller.ts:79]: Similar issue to line 70 above. Response payload shape is inconsistent

File: src/services/groups.service.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
+import { PrismaClient, Group } from '@prisma/client';
+import { CreateGroupDto } from '@dtos/group.dto';
+import { HttpException } from '@exceptions/HttpException';
+import { isEmpty } from '@utils/util';
+
+class GroupService {
+  public group = new PrismaClient().group;
```
Comment (anjayluh) [inline src/services/groups.service.ts:7]: Instantiating `PrismaClient` directly inside the service leads to multiple DB connections when used in multiple places. Consider having a shared Prisma instance (singleton).

File: src/services/groups.service.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+  public async findAllGroup(): Promise<Group[]> {
```
Comment (anjayluh) [inline src/services/groups.service.ts:9]: Use `findAllGroups()` for plural consistency

File: src/services/groups.service.ts [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    const allGroup: Group[] = await this.group.findMany();
+    return allGroup;
+  }
+
+  public async findGroupById(groupId: number): Promise<Group> {
+    if (isEmpty(groupId)) throw new HttpException(400, "groupId is empty");
```
Comment (anjayluh) [inline src/services/groups.service.ts:15]: It is a good thought to guard for empty ID. You also need to guard for `groupId` is NaN because isEmpty might not catch it. See comment on line 24 for example.

File: src/services/groups.service.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+    const findGroup = await this.group.findUnique({ where: { id: groupId }});
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
```
Comment (anjayluh) [inline src/services/groups.service.ts:18]: Status code `409 Conflict` is misleading here. For not found, the correct status is `404 Not Found`.

File: src/services/groups.service.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+    return findGroup;
+  }
+
+  public async createGroup(GroupData: CreateGroupDto): Promise<Group> {
+    if (isEmpty(GroupData)) throw new HttpException(400, "GroupData is empty");
```
Comment (anjayluh) [inline src/services/groups.service.ts:24]: Validation of data transfer object is shallow because it only checks for empty object thus excluding required fields.  This does not validate the actual fields of CreateGroupDto (e.g., required fields like name, description). So even if groupData has a key missing, this check passes — the object is not truly validated.

Proper approach with class-validator:

```
import { validateOrReject } from 'class-validator';
import { plainToInstance } from 'class-transformer';

const groupDataInstance = plainToInstance(CreateGroupDto, groupData);
await validateOrReject(groupDataInstance);
```

File: src/services/groups.service.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+  public async findGroupById(groupId: number): Promise<Group> {
+    if (isEmpty(groupId)) throw new HttpException(400, "groupId is empty");
+
+    const findGroup = await this.group.findUnique({ where: { id: groupId }});
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
+
+    return findGroup;
+  }
+
+  public async createGroup(GroupData: CreateGroupDto): Promise<Group> {
```
Comment (anjayluh) [inline src/services/groups.service.ts:23]: GroupData should be camelCase (`groupData`). In TypeScript/JavaScript, class names are PascalCase by convention `CreateGroupDto`, while variable or parameter names are camelCase (`groupData`). So here, the class is `CreateGroupDto`, and the parameter should be groupData for readability and to follow conventions

`public async createGroup(GroupData: CreateGroupDto)` becomes `public async createGroup(groupData: CreateGroupDto)`

File: src/services/groups.service.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+
+    const createGroupData: Group = await this.group.create({ data: { ...GroupData, }});
+    return createGroupData;
+  }
+
+  public async updateGroup(groupId: number, GroupData: CreateGroupDto): Promise<Group> {
+    if (isEmpty(GroupData)) throw new HttpException(400, "GroupData is empty");
+
+    const findGroup = await this.group.findUnique({ where: { id: groupId } });
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
```
Comment (anjayluh) [inline src/services/groups.service.ts:34]: Similar issue to line 18. Status code should be `404 Not Found`.

File: src/services/groups.service.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+    if (isEmpty(GroupData)) throw new HttpException(400, "GroupData is empty");
+
+    const createGroupData: Group = await this.group.create({ data: { ...GroupData, }});
+    return createGroupData;
+  }
+
+  public async updateGroup(groupId: number, GroupData: CreateGroupDto): Promise<Group> {
+    if (isEmpty(GroupData)) throw new HttpException(400, "GroupData is empty");
+
+    const findGroup = await this.group.findUnique({ where: { id: groupId } });
```
Comment (anjayluh) [inline src/services/groups.service.ts:33]: I appreciate the guard of using `findUnique` before update prevents Prisma throwing unhandled errors.

File: src/services/groups.service.ts [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
+
+    const updateGroupData = await this.group.update({ where: { id: groupId }, data: { ...GroupData, }});
+    return updateGroupData;
+  }
+
+  public async deleteGroup(groupId: number): Promise<Group> {
+    if (isEmpty(groupId)) throw new HttpException(400, "Group doesn't existId");
```
Comment (anjayluh) [inline src/services/groups.service.ts:41]: - The message "Group doesn't existId" is a typo and unclear. I suggest changing it to "Group ID is required".
- The issue mentioned on line 18 also applies here i.e. status code should be 404 instead of 409.

File: src/services/groups.service.ts [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+    const findGroup: Group = await this.group.findUnique({ where: { id: groupId } });
+    if (!findGroup) throw new HttpException(409, "Group doesn't exist");
+
+    const deleteGroupData = await this.group.delete({ where: { id: groupId } });
+    return deleteGroupData;
+  }
+
+  public async joinGroup(userId: number, groupId: number) {
+    if (isEmpty(groupId) || isEmpty(userId)) throw new HttpException(400, "Group doesn't existId");
```
Comment (anjayluh) [inline src/services/groups.service.ts:51]: - The error message is misleading because it implies the group doesn’t exist, but the problem could also be that the userId is missing.

Suggestion: separate the checks or clarify the message:

```
if (isEmpty(groupId)) throw new HttpException(400, "Group ID must be provided");
if (isEmpty(userId)) throw new HttpException(400, "User ID must be provided");
```
- isEmpty does not catch NaN. Number(req.params.id) could produce NaN, so consider adding a numeric check:

`if (isNaN(groupId)) throw new HttpException(400, "Group ID must be a valid number");`

Comment (anjayluh) [reply src/services/groups.service.ts:51]: I suggest you also check if group exists

```
const groupExists = await this.group.findUnique({
    where: { id: groupId },
    select: { id: true, users: { select: { user_id: true } } }
  });
  if (!groupExists) throw new HttpException(404, "Group not found");
```

File: src/services/groups.service.ts [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
```
Comment (anjayluh) [inline src/services/groups.service.ts:52]: There is no check for duplicates. If the user is already in the group, this will throw a Prisma constraint error. Consider checking their existence first:

  ```
const alreadyInGroup = groupExists.users.some(u => u.user_id === userId);
  if (alreadyInGroup) throw new HttpException(409, "User already in group");
```

Comment (anjayluh) [reply src/services/groups.service.ts:52]: Safety suggestion: If you plan to do multiple related updates (e.g., logging, notifications), consider wrapping in a transaction to prevent partial updates.

File: src/services/groups.service.ts [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+      },
+      select: {
+        users: true
+      }
+    });
+
+    return joinGroupData;
+  }
+
+  public async leaveGroup(userId: number, groupId: number) {
```
Comment (anjayluh) [inline src/services/groups.service.ts:72]: Similar to joinGroup above:

- Error message is misleading
- Consider splitting validation for clarity
- Add check if the user actually exists in the group before attempting deletion

File: src/services/users.service.ts [hunk 25]
```diff
@@ -52,6 +52,17 @@ class UserService {
     const deleteUserData = await this.users.delete({ where: { id: userId }, select: { email: true, id: true } });
     return deleteUserData;
   }
+
+  public async manageGroups(userId: number, data: any) {
+    if (isEmpty(data)) throw new HttpException(400, "Groups don't exist");
```
Comment (anjayluh) [inline src/services/users.service.ts:57]: The error message is misleading because data could be empty for reasons other than missing groups. `data: any` is too loose; it allows anything (string, number, object, array). I recommend you define a DTO for the expected shape of data (e.g., `ManageGroupsDto`) and validate it properly using class-validator. This ensures type safety and prevents runtime errors.

```
if (!data || !Array.isArray(data.groups) || data.groups.length === 0)
  throw new HttpException(400, "Groups is required");
```

File: src/services/users.service.ts [hunk 26]
```diff
@@ -52,6 +52,17 @@ class UserService {
     const deleteUserData = await this.users.delete({ where: { id: userId }, select: { email: true, id: true } });
     return deleteUserData;
   }
+
+  public async manageGroups(userId: number, data: any) {
```
Comment (anjayluh) [inline src/services/users.service.ts:56]: manageGroups is vague because it doesn’t specify what “manage” means (add, remove, update?).  Use a more explicit name like `updateUserGroups`. This improves readability and maintainability.

File: src/services/users.service.ts [hunk 27]
```diff
@@ -52,6 +52,17 @@ class UserService {
...
+    if (isEmpty(data)) throw new HttpException(400, "Groups don't exist");
+    
+    const groups = await this.users.update({
+      where: { id: userId },
+      data
+    })
+
+    return groups;
```
Comment (anjayluh) [inline src/services/users.service.ts:64]: For consistency with your other service methods, consider returning only the relevant fields instead of returning the raw Prisma result.

`return { id: ..., groups: ... };`

File: src/server.ts [hunk 28]
```diff
@@ -1,11 +1,12 @@
 import App from '@/app';
 import AuthRoute from '@routes/auth.route';
+import GroupsRoute from './routes/group.route';
 import IndexRoute from '@routes/index.route';
 import UsersRoute from '@routes/users.route';
 import validateEnv from '@utils/validateEnv';
 
 validateEnv();
```
Comment (anjayluh) [inline src/server.ts:8]: This is a good practice to ensure that all required environment variables are loaded and valid before starting the app. However, wrap in try/catch to log errors clearly and exit the process if validation fails to prevent the app from starting in a misconfigured state.

For example: 

```
try {
  validateEnv();
} catch (error) {
  console.error('Environment validation failed:', error);
  process.exit(1);
}

```

File: src/server.ts [hunk 29]
```diff
@@ -1,11 +1,12 @@
 import App from '@/app';
 import AuthRoute from '@routes/auth.route';
+import GroupsRoute from './routes/group.route';
```
Comment (anjayluh) [inline src/server.ts:3]: `GroupsRoute` is imported with a relative path, whereas others use alias (@routes/...). Unify paths for readability and maintainability:

File: src/server.ts [hunk 30]
```diff
@@ -1,11 +1,12 @@
...
 import IndexRoute from '@routes/index.route';
 import UsersRoute from '@routes/users.route';
 import validateEnv from '@utils/validateEnv';
 
 validateEnv();
 
-const app = new App([new IndexRoute(), new UsersRoute(), new AuthRoute()]);
+const app = new App([new IndexRoute(), new UsersRoute(), new AuthRoute(), new GroupsRoute()]);
```
Comment (anjayluh) [inline src/server.ts:10]: Passing an array of route instances to a central App class is clean and scalable.
Minor suggestion: Order of routes might matter if some routes have overlapping paths (/ vs /users). Typically, more specific routes come first.

File: src/server.ts [hunk 31]
```diff
@@ -1,11 +1,12 @@
...
 
 app.listen();
```
Comment (anjayluh) [inline src/server.ts:12]: Optional but helpful: Consider adding logging to confirm the server started and on which port. This is especially helpful in production or for debugging during development.

File: src/routes/group.route.ts [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+  public router = Router();
+  public groupsController = new GroupsController();
+
+  constructor() {
+    this.initializeRoutes();
+  }
+
+  private initializeRoutes() {
+    this.router.get(`${this.path}`, this.groupsController.getGroups);
+    this.router.get(`${this.path}/:id(\\d+)`, this.groupsController.getGroupById);
```
Comment (anjayluh) [inline src/routes/group.route.ts:18]: Consider ordering routes carefully such that more specific routes come before general ones in a scenario where paths might overlap.

File: src/routes/group.route.ts [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+    this.router.post(`${this.path}`, validationMiddleware(CreateGroupDto, 'body'), this.groupsController.createGroup);
+    this.router.put(`${this.path}/:id(\\d+)`, validationMiddleware(CreateGroupDto, 'body', true), this.groupsController.updateGroup);
+    this.router.delete(`${this.path}/:id(\\d+)`, this.groupsController.deleteGroup);
+    this.router.put(`${this.path}/join/:id(\\d+)`, this.groupsController.joinGroup);
+    this.router.put(`${this.path}/leave/:id(\\d+)`, this.groupsController.leaveGroup);
```
Comment (anjayluh) [inline src/routes/group.route.ts:23]: - Using verbs in path (join, leave) violates RESTful conventions.
- Method PUT is used, but adding/removing a membership is not idempotent on the group itself. Instead, it modifies a sub-resource (members).
- POST is more semantically appropriate for join (creating a membership) and DELETE for leave (removing membership)

The proper semantics are:

```
// Add a user to a group
POST   /groups/:id/members         -> joinGroup

// Remove a user from a group
DELETE /groups/:id/members/:userId -> leaveGroup
```
This ensures:
- Clearer semantics i.e. routes now reflect the resource being modified (members) instead of actions.
- HTTP methods match resource operations: POST for creation, DELETE for deletion.
- Aligns with RESTful best practices.

Standard HTTP Methods for CRUD include:

**GET → read, POST → create, PUT → update, DELETE → delete — standard CRUD operations.**

File: src/routes/group.route.ts [hunk 34]
```diff
@@ -0,0 +1,27 @@
+import { Router } from 'express';
+import GroupsController from '@controllers/groupe.controller';
```
Comment (anjayluh) [inline src/routes/group.route.ts:2]: As mentioned, unify spelling (groupe → groups) to maintain consistency with route names and class names.

File: src/controllers/users.controller.ts [hunk 35]
```diff
@@ -60,6 +60,18 @@ class UsersController {
       next(error);
     }
   };
+
+  public manageGroups = async (req: Request, res: Response, next: NextFunction) => {
```
Comment (anjayluh) [inline src/controllers/users.controller.ts:64]: Method name `manageGroups`  is vague as it does not clearly convey what the endpoint does (add/remove groups, assign roles, etc.). More descriptive naming like `updateUserGroups` would be clearer.

Use descriptive names that match the action in the response message field as well.

File: src/controllers/users.controller.ts [hunk 36]
```diff
@@ -60,6 +60,18 @@ class UsersController {
...
+    try {
+      const data = req.body;
+      const userId = Number(req.params.id);
```
Comment (anjayluh) [inline src/controllers/users.controller.ts:67]: Casting with Number() is acceptable, but consider validating the route param to ensure it’s numeric before calling the service.

File: src/controllers/users.controller.ts [hunk 37]
```diff
@@ -60,6 +60,18 @@ class UsersController {
       next(error);
     }
   };
+
+  public manageGroups = async (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const data = req.body;
```
Comment (anjayluh) [inline src/controllers/users.controller.ts:66]: Similar issue of data is typed as any and this bypasses TypeScript safety

File: src/controllers/users.controller.ts [hunk 38]
```diff
@@ -60,6 +60,18 @@ class UsersController {
...
+      const userId = Number(req.params.id);
+      const groupData = await this.userService.manageGroups(userId, data);
```
Comment (anjayluh) [inline src/controllers/users.controller.ts:68]: It is unsafe to pass raw data directly to the service because:

1. Service currently does not enforce type safety or validate the group IDs exist.
2. It could overwrite unrelated user fields in case data contains extra keys.

Extract the intended property before passing:

```
const groupIds = data.groupIds;
const groupData = await this.userService.manageGroups(userId, { groups: groupIds });
```


This makes the intention explicit and reduces risk of accidental data corruption.

File: src/controllers/users.controller.ts [hunk 39]
```diff
@@ -60,6 +60,18 @@ class UsersController {
...
+
+      res.status(200).json({ data: groupData, message: 'manageGroups' });
```
Comment (anjayluh) [inline src/controllers/users.controller.ts:70]: - `message: 'manageGroups'` is not descriptive. Use a clearer text like "User groups updated".

Safety check: groupData should only include safe fields (e.g., IDs or names), not raw Prisma objects that might leak sensitive user data.

File: src/prisma/schema.prisma [hunk 40]
```diff
@@ -11,7 +11,25 @@ generator client {
...
+  user     User  @relation(fields: [user_id], references: [id], onDelete: Cascade)
+  group_id Int
+  group    Group @relation(fields: [group_id], references: [id], onDelete: Cascade)
+
+  @@id([user_id, group_id])
+}
+
+model Group {
+  id    Int      @id @default(autoincrement())
+  name  String
```
Comment (anjayluh) [inline src/prisma/schema.prisma:31]: Enforce name uniqueness of group names to prevent accidental duplicate groups.

File: src/prisma/schema.prisma [hunk 41]
```diff
@@ -11,7 +11,25 @@ generator client {
 }
 
 model User {
-  id       Int    @id @default(autoincrement())
-  email    String @unique
+  id       Int      @id @default(autoincrement())
+  email    String   @unique
   password String
+  groups   Member[]
+}
```
Comment (anjayluh) [inline src/prisma/schema.prisma:18]: Add createdAt / updatedAt timestamps for auditing

```
createdAt DateTime @default(now())
updatedAt DateTime @updatedAt
```

File: src/prisma/schema.prisma [hunk 42]
```diff
@@ -11,7 +11,25 @@ generator client {
...
+
+model Member {
+  user_id  Int
+  user     User  @relation(fields: [user_id], references: [id], onDelete: Cascade)
+  group_id Int
+  group    Group @relation(fields: [group_id], references: [id], onDelete: Cascade)
+
+  @@id([user_id, group_id])
+}
```
Comment (anjayluh) [inline src/prisma/schema.prisma:27]: Similar comment as on line 18 above to add createdAt / updatedAt timestamps for auditing

File: src/prisma/schema.prisma [hunk 43]
```diff
@@ -11,7 +11,25 @@ generator client {
...
+
+model Group {
+  id    Int      @id @default(autoincrement())
+  name  String
+  users Member[]
+
+  @@index([name])
 }
```
Comment (anjayluh) [inline src/prisma/schema.prisma:35]: Similar comment as on line 18 above to add createdAt / updatedAt timestamps for auditing

**Review by anjayluh — COMMENTED** [review]
Hello @ben-at-soul-refiner 

Here are the things we need to work on to get this ready for deployment.

1. We need to automatically verify that the code behaves as expected by adding comprehensive test suites using Jest and Supertest. Forexample;
        -   **Controllers/Service:** Unit tests for `GroupsService` methods (create, find, update, delete, join, leave) mocking the Prisma client.
        -   **API Endpoints:** Integration tests for all REST routes (`GET /groups`, `POST /groups/:id/members`, etc.) verifying status codes, response bodies, and database state changes.
        -   **Edge Cases:** Tests for error conditions (e.g., user not found, group not found, duplicate membership).

2. We need to ensure that no endpoint processes unvalidated data. The  DTO classes are defined but not consistently validated. Simply casting `req.body` is unsafe and bypasses validation decorators. 
We will need to implement a validation pipe (or use `class-validator`'s `validateOrReject`) in the global middleware to automatically validate all incoming requests against the DTOs. 

3. It is important to have a valuable git structure for collaboration, debugging, and understanding why changes were made. That way, the reviewer saves time on understanding the context; thus, reducing back and forth, and increasing the quality and speed of pushing to. production. We can use something like so:
    -   **Branch Naming:** We will need to use descriptive names like `feature/groups-module` or `fix/groups-typo`,  avoiding generic names like `pullrequest`. 
    -   **Commit History:**  . To create a clear, bisectable history. We should consider using [Conventional Commits](https://www.conventionalcommits.org/), and use atomic, granular commits (e.g., `feat(groups): add controller and routes`, `feat(groups): implement join/leave service logic`)
    - **PR Description**
    -   **Issue:** We could give the description a more critical context. We can take up a format like so:
    -   **Request:** In the future, always include:
        -   **Objective:** The purpose of this change.
        -   **Summary of Changes:** A high-level list of what was added/modified.
        -   **Testing Instructions:** Steps for a reviewer to test the functionality manually.
        -   **Linked Issues:** `Closes #123`.

4.  **RESTful API Design**
    -   **Issue:** The routes `POST /groups/join/:id` and `POST /groups/leave/:id` are not RESTful. They are RPC-like and don't leverage HTTP semantics correctly.
    -   **Request:** We could refactor the membership endpoints to act on the `members` sub-resource.
        -   **Join:** `POST /groups/:groupId/members` (Body: `{ userId: string }`)
        -   **Leave:** `DELETE /groups/:groupId/members/:userId`
    -   **Rationale:** This aligns with standard REST conventions, making the API intuitive and predictable for consumers.

5.  **Consistent Error Handling & Response Formatting** 
    -   **Issue:** Response payloads are inconsistent. Some endpoints return raw objects, others return a `{ data }` wrapper. Error handling is duplicated and inconsistent.
    -   **Request:**
        -  We will need to define a standard response format (e.g., `{ success: boolean, data: T, message?: string }`) and use it across all endpoints.
        -   Implement a centralized error handling middleware to catch and format all errors (Prisma errors, validation errors, custom `HttpException`s) into a consistent structure.

6.  **Adhere to DRY Principle**
    -   **Issue:** Repetitive checks for resource existence (e.g., checking if a user or group exists) are present in multiple methods.
    -   **Request:** We need to  refactor these checks into reusable, shared utility functions or within a custom Prisma middleware. This reduces code duplication and makes logic easier to test and maintain.

7. **Code Consistency**
    -   **Naming:** We should ensure variable naming follows `camelCase` consistently (`groupData`, not `GroupData`).
    -   **File Names:** We should correct the typo in `groupe.controller.ts` to `group.controller.ts`.


I have also made some comments below for you.

### **Final Verdict**

This is a very strong submission that demonstrates core technical competency. The architectural choices are sound. The requested changes focus on elevating the work from a functioning feature to a production-ready, maintainable, and collaborative piece of software.

Once these changes are addressed, we should be good to go. I am happy to provide further clarification on any of these points.

**Well done.**
